cartographer/pose_graph/pose_graph_data.cc

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/pose_graph/pose_graph_data.h"
+
+#include "cartographer/common/make_unique.h"
+#include "cartographer/pose_graph/constraint/acceleration_constraint_3d.h"
+#include "cartographer/pose_graph/constraint/interpolated_relative_pose_constraint_3d.h"
+#include "cartographer/pose_graph/constraint/relative_pose_constraint_2d.h"
+#include "cartographer/pose_graph/constraint/relative_pose_constraint_3d.h"
+#include "cartographer/pose_graph/constraint/rotation_constraint_3d.h"
+
+namespace cartographer {
+namespace pose_graph {
+namespace {
+
+std::unique_ptr<Constraint> CreateConstraint(
+    const proto::Constraint& constraint) {
+  const auto& id = constraint.id();
+  const auto& loss = constraint.loss_function();
+  const auto& cost = constraint.cost_function();
+  switch (cost.type_case()) {
+    case (proto::CostFunction::kRelativePose2D):
+      return common::make_unique<RelativePoseConstraint2D>(
```

Comment:
I like this pattern much better than the one used in the nodes. What do you think about having a CreateNode() function as well?

Reply:
How would it look like? Nodes are not really stored by keeping a ptr to the base class.

Reply:
Oh right, my bad. Carry on :-)